Podman6: Remove slirp

lsm5 opened the pull request:
<!--
Thanks for sending a pull request!

For more detailed information, please review our contributing guidelines:
https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests
-->

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:
<!-- Use [x] to mark as done, or click the checkbox after opening PR -->

- [ ] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [ ] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [ ] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [ ] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [ ] All commits pass `make validatepr` (format/lint checks)
- [ ] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

<!--
Write `None` if there are no user-facing changes, otherwise enter your release note below.
Include "action required" if users need to take action when upgrading.
-->

```release-note
Slirp support has been removed.
```

lsm5 committed:
test: Remove slirp4netns-specific test cases

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Remove Slirp network mode constant and error on usage

Using --network=slirp4netns now returns a hard error directing
users to pasta and `podman system migrate`.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Remove slirp4netns setup functions (preserve RLK)

Remove all slirp4netns-specific setup and helper functions while
preserving the RootlessRLK port mapping functions that are still
used by pasta and bridge networking.

Rename networking_slirp4netns.go to networking_rootlessport.go.
Convert the slirp4netns compose test to use pasta instead.
Remove rootlessSlirpSyncR/W fields from container struct and
getSlirp4netnsIP stub from FreeBSD networking.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Remove slirp4netns from CLI and completions

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Remove slirp4netns from documentation

Update all documentation files to remove slirp4netns references
and update to pasta as the only rootless networking backend.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Remove unused slirp4netns imports and Slirp4NetNS field

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
FreeBSD: add lint exclusion for rootlessPortSync fields

The rootlessPortSyncR/W fields in the Container struct are only
used on Linux but defined in the shared container.go. Add a linter
exclusion so they don't trigger unused warnings on FreeBSD.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Clean up outdated slirp4netns references in comments

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

mtrmac commented on docs/source/locale/ja/LC_MESSAGES/markdown.po:
```diff
@@ -3263,23 +3263,12 @@ msgstr ""
 #: ../../source/markdown/podman-pod-create.1.md:410
 #: ../../source/markdown/podman-run.1.md:1214
 msgid ""
-"**slirp4netns[:OPTIONS,...]**: use **slirp4netns**(1) to create a user "
```
These are incorrect, `msgid` and `msgstr` are ~key-value pairs.

More importantly, this file should be auto-generated from the source text using … some process.

I _guess_ that should be https://www.sphinx-doc.org/en/master/usage/advanced/intl.html but I don’t see anything to suggest we have that automated.

mtrmac reviewed (commented):
By no means a full review, I just checked whether this can be quickly merged and stopped at the first blocker.

lsm5 committed:
Add slirp4netns to pasta migration

Add slirp4netns to pasta migration in `podman system migrate`. When
run, it changes the NetMode from slirp4netns to pasta and removes
any slirp4netns network options. Warns about options that cannot be
mapped to pasta (e.g. cidr, allow_host_loopback). The config is
persisted via RewriteContainerConfig.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Set pasta as default when default_rootless_network_cmd is unset

Previously, when default_rootless_network_cmd was empty, podman
would fall back to slirp4netns. Change this to default to pasta
instead. Remove the now-unused slirp4netns import.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
RPM/TMT: drop slirp4netns dependency

slirp4netns is no longer used as a rootless networking backend,
pasta is the only option. Remove the package requirement from
rpm/podman.spec and the TMT test setup.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 committed:
Fix pasta_opts compose test to verify MTU option

Use podman exec to check ip link MTU instead of ncat-based host
connectivity test which does not actually validate pasta options.

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 commented:
I think everything's addressed so far. Moved the migration to system migrate. PTAL.

Luap99 commented on test/compose/pasta_opts/docker-compose.yml:
```diff
@@ -0,0 +1,5 @@
+  alpine:
+    image: alpine
+    network_mode: "pasta:--ipv4-only,-a,10.0.2.0,-n,24,-g,10.0.2.2,--dns-forward,10.0.2.3,--mtu,1280,--no-ndp,--no-dhcpv6,--no-dhcp"
+    command: sh -c "echo teststring | nc 10.0.2.2 5001"
```
I need to check this, the gateway should not allow a connecting by default to the host, so I think this test might be broken and not catch failures.

Anyhow not that important overall and could be tracked as follow up.

Luap99 replied:
And now that I am at the last commit I see you fixed up the test, except the test as is does not validate any options that we parsed as the host.containers.internal thing is done by default.

I would suggest we drop this ncat thing altogether, just run sleep here and then as test to `podman exec $containername ip link` and then match output contains `mtu 1280` to know the mtu was set

Luap99 commented on pkg/domain/infra/abi/network.go:
```diff
@@ -180,7 +179,7 @@ func (ic *ContainerEngine) NetworkRm(ctx context.Context, namesOrIds []string, o
 }
 
 func (ic *ContainerEngine) NetworkCreate(_ context.Context, network types.Network, createOptions *types.NetworkCreateOptions) (*types.Network, error) {
-	if slices.Contains([]string{"none", "host", "bridge", "private", slirp4netns.BinaryName, pasta.BinaryName, "container", "ns", "default"}, network.Name) {
```
So here I am wondering if we should keep rejecting slirp4netns as name, it is not like allowing that would cause an issue but it could certainly cause confusing if a users has a mixed system with podman 6 and some still on 5 were they result in totally different behaviours.

No strong opinion

Luap99 commented on docs/source/markdown/podman-create.1.md.in:
```diff
@@ -499,13 +499,13 @@ be installed. The shadow-utils package must include the newuidmap and newgidmap
 In order for users to run rootless, there must be an entry for their username in /etc/subuid and /etc/subgid which lists the UIDs for their user namespace.
 
-Rootless Podman works better if the fuse-overlayfs and slirp4netns packages are installed.
+Rootless Podman works better if the fuse-overlayfs package is installed.
```
Not for this Pr but we should create an issue this is just outdated, we no longer need fuse

Luap99 commented on docs/source/markdown/podman-network.1.md:
```diff
@@ -35,10 +35,7 @@ The default bridge network (called `podman`) uses 10.88.0.0/16 as a subnet. When
-
-### Slirp4netns
-Slirp4netns uses 10.0.2.0/24 for its default network. This can also be changed in **[containers.conf(5)](https://github.com/containers/container-libs/blob/main/common/docs/containers.conf.5.md)** but under the `[engine]` section. Use the `network_cmd_options` key and add `["cidr=X.X.X.X/24"]` as a value. Note that slirp4netns needs a network prefix size between 1 and 25. This option accepts an ar ...
+Pasta is the default rootless networking tool. This can be configured in **[containers.conf(5)](https://github.com/containers/container-libs/blob/main/common/docs/containers.conf.5.md)** under the `[network]` section with `default_rootless_network_cmd`.
```
This can be configured seems misleading, we support exactly one so? Maybe just remove the reference to the default_rootless_network_cmd config setting.

Luap99 commented on docs/tutorials/socket_activation.md:
```diff
@@ -264,9 +264,9 @@ container then runs with less privileges.
-When using rootless Podman, network traffic is normally passed through slirp4netns. This comes with
-a performance penalty. Fortunately, communication over the socket-activated socket does not pass through
-slirp4netns so it has the same performance characteristics as the normal network on the host.
+When using rootless Podman, network traffic is normally passed through pasta. Fortunately, communication
```
keep the ` This comes with a performance penalty`, while pasta is supposed to be faster it will not be as fast as socket activation in this context here

Luap99 commented on contrib/cirrus/logcollector.sh:
```diff
@@ -47,7 +47,7 @@ case $1 in
                     podman
                     skopeo
-                    slirp4netns
+
```
nit: That is not really a comment (as claimed in the commit but I do not mind much) but the empty line should be removed here

Luap99 commented on libpod/networking_rootlessport.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 	"github.com/containers/podman/v6/pkg/errorhandling"
 	"github.com/sirupsen/logrus"
+
```
that empty line does not seem needed

lsm5 committed:
Remove misleading configurable reference in podman-network docs

Signed-off-by: Lokesh Mandvekar <lsm5@redhat.com>

lsm5 commented:
Converting back to draft while I wait for tests to pass.

Addressed all comments and added `slirp4netns` back into the rejected network name list which sounds safer to me.

lsm5 commented:
The testing-farm job is failing on a dnf repo issue which has been reported upstream. We can ignore it here. I've triggered a rerun of the machine test, but anyway, should be good for another round.

@Luap99 @mheon @containers/podman-maintainers PTAL

Luap99 commented on test/compose/pasta_opts/setup.sh:
nit: this file can be removed if it does nothing

lsm5 replied:
removed

Luap99 commented on test/compose/pasta_opts/teardown.sh:
nit: same, this file can be removed if it does nothing

lsm5 replied:
removed

Luap99 commented on libpod/networking_rootlessport.go:
```diff
@@ -11,10 +11,7 @@ import (
-	// functionality (SetupRootlessPortMappingViaRLK and GetRootlessPortChildIP),
-	// NOT for the deprecated slirp4netns network backend which was removed in Podman 6.
-	"go.podman.io/common/libnetwork/slirp4netns"
+	"go.podman.io/common/libnetwork/slirp4netns" // RootlessKit port mapping only, not the removed slirp4netns backend
```
That did not actually solve my newline comment, first you still have the newline.

Commit e8ad9f6c50c383b3c9d307f78566d09264a1570f adds a newline above as part of a commit where it really does not belong. For git history reasons we should not have such "random" newline changes.
Sure not like I care to git blame to the import statements but at least right now for things like possible merge conflicts it still matters.

So really these changes should be squashed for a proper clean history.

lsm5 replied:
squashed

Luap99 commented on docs/tutorials/socket_activation.md:
```diff
@@ -264,9 +264,9 @@ container then runs with less privileges.
-When using rootless Podman, network traffic is normally passed through pasta. Fortunately, communication
-over the socket-activated socket does not pass through pasta so it has the same performance characteristics
-as the normal network on the host.
+When using rootless Podman, network traffic is normally passed through pasta. This comes with
+a performance penalty. Fortunately, communication over the socket-activated socket does not pass through
+pasta so it has the same performance characteristics as the normal network on the host.
 
```
same with this commit now, it should just be sqashed, removing then re-adding makes reviewing much harder

lsm5 replied:
squashed

Luap99 reviewed (approved):
LGTM

lsm5 commented:
@mheon @containers/podman-maintainers PTAL

Honny1 reviewed (commented):
LGTM
